scripts/datastore_upgrade.py

```diff
@@ -0,0 +1,324 @@
+
+      for entity in entities:
+        process_entity(entity, datastore, ds_distributed)
+
+      last_key = entities[-1].keys()[0]
+      entities_checked += len(entities)
+
+      if time.time() > last_logged + LOG_PROGRESS_FREQUENCY:
+        logger.info("Checked {} entities".format(entities_checked))
+        last_logged = time.time()
+
+      if len(entities) < BATCH_SIZE:
+        break
+
+    except datastore_errors.Error as error:
+      logger.error("Error getting a batch of entities: {}".format(error))
+    except AppScaleDBConnectionError as connection_error:
+      logger.error("Error getting a batch of entities: {}".format(connection_error))
+
+def get_entity_batch(last_key, datastore, batch_size):
+  """ Gets a batch of entities to operate on.
+  Args:
+    last_key: The last key from a previous query.
+    datastore: A reference to the batch datastore interface.
+    batch_size: The max number of entities retrieved in the datastore request.
+  Returns:
+    A list of entities with a limit of batch_size.
+  """
+  return datastore.range_query(APP_ENTITY_TABLE, APP_ENTITY_SCHEMA, last_key,
+    "", batch_size, start_inclusive=False)
+
+def process_entity(entity, datastore, ds_distributed):
+  """ Processes an entity by updating it if necessary and removing tombstones.
+  Args:
+    entity: The entity to process.
+    datastore: A reference to the batch datastore interface.
+    ds_distributed: A reference to the distributed datastore.
+  Returns:
+    True on success, False otherwise.
+  """
+  logger.debug("Process entity {}".format(str(entity)))
+  key = entity.keys()[0]
+  one_entity = entity[key][APP_ENTITY_SCHEMA[0]]
+  version = entity[key][APP_ENTITY_SCHEMA[1]]
+  if one_entity == datastore_server.TOMBSTONE:
+    return delete_entity_from_table(key, datastore)
+
+  entity_proto = entity_pb.EntityProto(one_entity)
+  app_id = entity_proto.key().app()
+  validated_entity = ds_distributed.validated_result(app_id, entity)
+
+  if not validated_entity:
+    return True
+
+  if not (one_entity == validated_entity[key][APP_ENTITY_SCHEMA[0]] and version == validated_entity[key][APP_ENTITY_SCHEMA[1]]):
+    return update_entity_in_table(key, validated_entity, datastore)
+  return True
+
+def update_entity_in_table(key, validated_entity, datastore):
+  """ Updates the APP_ENTITY_TABLE with the valid entity.
+  Args:
+    key: A str representing the row key to update in the table.
+    validated_entity: A validated entity which needs to be updated in place of
+    the current entity.
+    datastore: A reference to the batch datastore interface.
+  Returns:
+    True: if updating entity was successful.
+    False: if an exception was encountered while updating the entity.
+  """
+  cell_values_to_update = {}
+  cell_values_to_update[key] = {
+    APP_ENTITY_SCHEMA[0]: validated_entity[key][APP_ENTITY_SCHEMA[0]],
+    APP_ENTITY_SCHEMA[1]: validated_entity[key][APP_ENTITY_SCHEMA[1]]
+  }
```

Comment:
You don't need to create a new dict here. You can just pass `validated_entity[key]` to `batch_put_entity` instead of `cell_values_to_update`.